### PR TITLE
Boundary conditions refactor and addition of non-characteristic slip and no-slip walls

### PR DESCRIPTION
**Pull request by wilfonba**
This takes care of #187 and part of #185 . It moves all of the boundary conditions stuff that isn't characteristic to a new module `m_boundary_conditions` that is called from `m_rhs`. I don't think moving CBCs to this module makes much sense since those are utilized in `m_rhs` in such a different way. All old test cases pass and 6 new test cases have been added for the new boundary conditions. I haven't yet verified the new BCs, but wanted this code to be available to @anshgupta1234 and @haochey so they had an idea what things will look like soon. Also, if someone could take a careful look at my changes the the `m_checker` modules to make sure I didn't mess up any of the checks, particularly when cylindrical coordinates are being used, that'd be greatly appreciated.

### CODE REVIEW

**Comment by sbryngelson** [comment]
Lines of code in simulation/ doubled here, adding 900 lines for an 1800-line long m_boundary_conditions. How did this happen/is it needed? It seems like much could be refactored there.

**Comment by wilfonba** [comment]
~650 of the additional lines are a result of the new boundary conditions I added, ~60 lines are from adding missing QBMM assignments to the symmetry case for `bc_x%beg` and `bc_x%end`, and the rest is just added white space and a change in the logic. The logic in the `s_populate_conservative_variables_buffers` is more or less repeated in all of the subroutines for each individual BC because this makes calls to those subroutines cleaner.

**Comment by sbryngelson** [comment]
@wilfonba looks like this is mostly good to go - rectify the merge conflicts and then do the minor refactor I believe you mentioned earlier today with `qbmm`. I don't want to actually merge this one until the GPU check is fixed.

**Comment by sbryngelson** [comment]
> I am wondering if we can use two-digit numbers (>-100) to name normal boundaries and three-digit numbers (<-100) to name characteristic boundaries. This approach helps avoid confusion between newly added boundary conditions and existing ones, and also distinguishes between normal and characteristic boundaries for coding logics in cases like "if (bc_x < -100) then ...".

@haochey good thinking. This seems reasonable to my eye. I'll let @wilfonba tell us if there's anything inherently wrong/problematic/annoying about doing this.

**Comment by wilfonba** [comment]
@sbryngelson This would make the code easier to read. The only down side is that it breaks every input file ever made. I'm not sure how beat to communicate with all relevant parties the change has been made to avoid frustrations.

**Comment by sbryngelson** [comment]
> @sbryngelson This would make the code easier to read. The only down side is that it breaks every input file ever made. I'm not sure how beat to communicate with all relevant parties the change has been made to avoid frustrations.

This is true. For now, let's stick to a solution that doesn't break old input files. This is an easy change later if appropriate.

**Comment by wilfonba** [comment]
@sbryngelson @haochey Might it be best to rename -5 to wall/solid-wall, and than have an if condition in the code that selectes no-slip wall if the problem is viscous and selects slip wall if the problem isn't? This would maintain all existing boundary conditions, including the seeimingly undocumented -13.

**Comment by sbryngelson** [comment]
> @sbryngelson @haochey Might it be best to rename -5 to wall/solid-wall, and than have an if condition in the code that selectes no-slip wall if the problem is viscous and selects slip wall if the problem isn't? This would maintain all existing boundary conditions, including the seeimingly undocumented -13.

Better would be to abort the case via `m_checker` if viscous + slip or inviscid + no-slip are chosen.

**Comment by wilfonba** [comment]
As long as this passes the tests this time, @haochey is going to make a PR to my fork that I'll then merge since this branch already has the logic for the additional boundary condition. Once he or I resolve merge conflicts and I add @haochey or @anshgupta1234 simple inflow condition (it's working well enough for them now), this should be ready to go.

File: src/simulation/m_cbc.fpp [hunk 1]
```diff
@@ -418,19 +421,115 @@ contains
...
                 ! ==================================================================
 
                 ! Computing CBC2 Coefficients ======================================
             elseif (weno_order == 3) then
 
-                fd_coef_${XYZ}$(:, cbc_loc_in) = 0d0
-                fd_coef_${XYZ}$(0, cbc_loc_in) = -6d0/(3d0*ds(0) + 2d0*ds(1) - ds(2))
-                fd_coef_${XYZ}$(1, cbc_loc_in) = -4d0*fd_coef_${XYZ}$(0, cbc_loc_in)/3d0
-                fd_coef_${XYZ}$(2, cbc_loc_in) = fd_coef_${XYZ}$(0, cbc_loc_in)/3d0
+                fd_coef_${XYZ}$(:, 1, cbc_loc_in) = 0d0
```
Comment (sbryngelson) [inline src/simulation/m_cbc.fpp:481]: please add comments to a paper/link for where these coefficients come from

File: src/simulation/m_cbc.fpp [hunk 2]
```diff
@@ -418,19 +421,115 @@ contains
...
+                    if (m*n*p > 0) then
+                    ! Trv2
+                        fd_coef_${XYZ}$(:, 3, cbc_loc_in) = 0d0
+                        fd_coef_${XYZ}$(0, 3, cbc_loc_in) = -1d0/(dz(0) + dz(1))
+                        fd_coef_${XYZ}$(1, 3, cbc_loc_in) = 0d0
+                        fd_coef_${XYZ}$(2, 3, cbc_loc_in) = -fd_coef_${XYZ}$(0, 3, cbc_loc_in)
+                    end if
+                
+                elseif (cbc_dir_in == 3) then
+                    ! Trv1
```
Comment (sbryngelson) [inline src/simulation/m_cbc.fpp:519]: trv1?

File: src/simulation/m_cbc.fpp [hunk 3]
```diff
@@ -418,19 +421,115 @@ contains
...
+                    fd_coef_${XYZ}$(:, 2, cbc_loc_in) = 0d0
+                    fd_coef_${XYZ}$(0, 2, cbc_loc_in) = -1d0/(dy(0) + dy(1))
+                    fd_coef_${XYZ}$(1, 2, cbc_loc_in) = 0d0
+                    fd_coef_${XYZ}$(2, 2, cbc_loc_in) = -fd_coef_${XYZ}$(0, 2, cbc_loc_in)
+
+                    if (m*n*p > 0) then
+                    ! Trv2
```
Comment (sbryngelson) [inline src/simulation/m_cbc.fpp:526]: ?

File: src/simulation/m_cbc.fpp [hunk 4]
```diff
@@ -418,19 +421,115 @@ contains
...
+                    end if
+                
+                elseif (cbc_dir_in == 3) then
+                    ! Trv1
+                    fd_coef_${XYZ}$(:, 2, cbc_loc_in) = 0d0
+                    fd_coef_${XYZ}$(0, 2, cbc_loc_in) = -1d0/(dy(0) + dy(1))
+                    fd_coef_${XYZ}$(1, 2, cbc_loc_in) = 0d0
+                    fd_coef_${XYZ}$(2, 2, cbc_loc_in) = -fd_coef_${XYZ}$(0, 2, cbc_loc_in)
+
+                    if (m*n*p > 0) then
```
Comment (sbryngelson) [inline src/simulation/m_cbc.fpp:525]: what does this condition mean? please add more comments
